SignalServiceKit/src/Messages/OWSMessageDecrypter.m

```diff
@@ -434,89 +435,137 @@ - (void)decryptUnidentifiedSender:(SSKProtoEnvelope *)envelope
-
-            if (error || !decryptResult) {
-                if ([error.domain isEqualToString:@"SignalMetadataKit.SMKSecretSessionCipherError"]
-                    && error.code == SMKSecretSessionCipherErrorSelfSentMessage) {
-                    // Self-sent messages can be safely discarded.
-                    return failureBlock(error);
-                }
-
-                OWSFailDebug(@"Could not decrypt UD message: %@", error);
-                error = EnsureDecryptError(error, @"Could not decrypt UD message");
+        NSError *decryptError;
+        SMKDecryptResult *_Nullable decryptResult =
+            [cipher throwswrapped_decryptMessageWithCertificateValidator:certificateValidator
+                                                          cipherTextData:encryptedData
+                                                               timestamp:serverTimestamp
+                                                        localRecipientId:localRecipientId
+                                                           localDeviceId:localDeviceId
+                                                         protocolContext:transaction
+                                                                   error:&decryptError];
+
+        if (!decryptResult) {
+            if (!decryptError) {
+                OWSFailDebug(@"Caller should provide specific error");
+                NSError *error = OWSErrorWithCodeDescription(
+                    OWSErrorCodeFailedToDecryptUDMessage, @"Could not decrypt UD message");
                 return failureBlock(error);
             }
 
-            if (decryptResult.messageType == SMKMessageTypePrekey) {
-                [TSPreKeyManager checkPreKeys];
+            // Decrypt Failure Part 1: Unwrap failure details
+
+            NSError *_Nullable underlyingError;
+            SSKProtoEnvelope *_Nullable identifiedEnvelope;
+
+            if (![decryptError.domain isEqualToString:@"SignalMetadataKit.SecretSessionKnownSenderError"]) {
+                underlyingError = decryptError;
+                identifiedEnvelope = envelope;
+            } else {
+                underlyingError = decryptError.userInfo[NSUnderlyingErrorKey];
+
+                NSString *senderRecipientId
+                    = decryptError.userInfo[SecretSessionKnownSenderError.kSenderRecipientIdKey];
+                OWSAssert(senderRecipientId);
+
+                NSNumber *senderDeviceId = decryptError.userInfo[SecretSessionKnownSenderError.kSenderDeviceIdKey];
+                OWSAssert(senderDeviceId);
+
+                SSKProtoEnvelopeBuilder *identifiedEnvelopeBuilder = envelope.asBuilder;
+                identifiedEnvelopeBuilder.source = senderRecipientId;
+                identifiedEnvelopeBuilder.sourceDevice = senderDeviceId.unsignedIntValue;
+                NSError *identifiedEnvelopeBuilderError;
+
+                identifiedEnvelope = [identifiedEnvelopeBuilder buildAndReturnError:&identifiedEnvelopeBuilderError];
```

Comment:
I don't see any checking of `identifiedEnvelopeBuilderError`.  At the very least we should log it if it's non-nil.

Reply:
Note that we *do* `OWSAssert(identifiedEnvelope)`. 

It would probably be a bug if we returned both an error and a result, but as I understand it, the recommended approach is to only consult the error if and only if the result was not successful.

But, I'm happy to add the logging you've requested.